File: src/main/java/indi/lt/serialtool/controller/SerialReceiveCtrl.java
package indi.lt.serialtool.controller;

import com.fazecast.jSerialComm.SerialPort;
import github.nonoas.jfx.flat.ui.concurrent.TaskHandler;
import indi.lt.serialtool.ConfigManager;
import indi.lt.serialtool.component.InlineCssRegexHighlighter;
import indi.lt.serialtool.component.PromptInlineCssTextArea;
import indi.lt.serialtool.component.SerialToggleButton;
import indi.lt.serialtool.service.SerialReadService;
import javafx.fxml.FXML;
import javafx.fxml.Initializable;
import javafx.scene.control.CheckBox;
import javafx.scene.control.ComboBox;
import javafx.scene.control.Label;
import javafx.scene.control.MenuBar;
import javafx.scene.control.TextField;
import javafx.scene.layout.BorderPane;
import javafx.stage.FileChooser;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.net.URL;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.ResourceBundle;

/**
 * @author Nonoas
 * @date 2025/8/22
 * @since 1.0.0
 */
public class SerialReceiveCtrl implements Initializable {
    private final Logger LOG = LogManager.getLogger(SerialController.class);

    @FXML
    private BorderPane rootPane;

    @FXML
    private MenuBar menuBar;

    @FXML
    private Label lbSerialName;

    @FXML
    private ComboBox<String> cbSerialList;

    @FXML
    private ComboBox<Integer> cbBautRateList;

    @FXML
    private PromptInlineCssTextArea textAreaOrigin;

    @FXML
    private SerialToggleButton btnOpenSerial;

    @FXML
    private CheckBox cbTimeDisplay;

    @FXML
    private TextField tfKeyWord;

    private SerialReadService serialReadService;

    private SerialPort comPort;

    private InlineCssRegexHighlighter highlighter;

    private String keyLastSerial;

    @Override
    public void initialize(URL url, ResourceBundle resourceBundle) {
        initBautRateList();
        registerSerialEvent();
    }

    private void registerSerialEvent() {
        initOpenSerialButtonAction();
        initBautRateComboBoxAction();
    }

    /**
     * 恢复自动滚动
     */
    @FXML
    private void restoreScrolling() {
        textAreaOrigin.setAutoScroll(true);
    }

    @FXML
    private void clearLogs() {
        textAreaOrigin.getArea().clear();
    }

    @FXML
    private void saveOriginLogs() {
        LOG.info("saveOriginLogs");

        // 1. 获取文本
        String content = textAreaOrigin.getText();
        if (content == null || content.isEmpty()) {
            LOG.info("没有日志内容可保存");
            return;
        }

        // 2. 选择保存文件
        FileChooser fileChooser = new FileChooser();
        fileChooser.setTitle("保存日志文件");
        fileChooser.getExtensionFilters().add(
                new FileChooser.ExtensionFilter("文本文件", "*.txt")
        );
        fileChooser.getExtensionFilters().add(
                new FileChooser.ExtensionFilter("所有文件", "*.*")
        );

        File file = fileChooser.showSaveDialog(textAreaOrigin.getScene().getWindow());
        if (file == null) {
            return; // 用户取消
        }

        // 3. 写入文件
        try (FileWriter writer = new FileWriter(file, false)) {
            writer.write(content);
            LOG.info("日志已保存到: " + file.getAbsolutePath());
        } catch (IOException e) {
            LOG.error("保存日志失败", e);
        }
    }


    @FXML
    private void saveFilterLogs() {
        LOG.info("saveFilterLogs");
        //textAreaOrigin.getArea().clear();
    }


    private void openSelectSerial(ComboBox<String> f_cbSerialList, ComboBox<Integer> f_cbBautRateList) {
        int selectedIndex = f_cbSerialList.getSelectionModel().getSelectedIndex();
        if (selectedIndex < 0) {
            LOG.info("请先选择一个串口设备");
            return;
        }

        SerialPort[] ports = SerialPort.getCommPorts();
        if (ports.length == 0) {
            LOG.info("未检测到串口设备");
            return;
        }
        //        Integer bautrate = cbBautRateList.getSelectionModel().getSelectedItem();
        int bautrate = Integer.parseInt(String.valueOf(f_cbBautRateList.getSelectionModel().getSelectedItem()));
        LOG.info("此时的波特率:" + bautrate);

        comPort = ports[selectedIndex];
        try {
            comPort.setComPortParameters(bautrate, 8, SerialPort.ONE_STOP_BIT, SerialPort.NO_PARITY);
            comPort.setComPortTimeouts(SerialPort.TIMEOUT_READ_SEMI_BLOCKING, 0, 0);
        } catch (Exception e) {
            LOG.error("设置波特率时发生异常", e);
            return;
        }

        if (!comPort.openPort()) {
            LOG.error("串口打开失败");
            return;
        }

        highlighter = new InlineCssRegexHighlighter(textAreaOrigin);
        highlighter.patternTextProperty().bind(tfKeyWord.textProperty());
        // 创建并启动Service
        this.serialReadService = new SerialReadService(comPort, textAreaOrigin,
                cbTimeDisplay,
                () -> highlighter.schedule());
        serialReadService.start();
    }

    public void initSerialComboBoxAction() {
        // 1. 展开下拉框时刷新列表（保留，但刷新逻辑已优化）
        cbSerialList.setOnShowing(event -> refreshSerialList());

        cbSerialList.getSelectionModel().selectedItemProperty().addListener((observableValue, oldVal, newVal) -> {
            if (null != newVal && null != oldVal) {
                LOG.debug(oldVal + "->" + newVal);
                onSerialPortClicked();
            }
        });

        List<String> serialList = new ArrayList<>();
        for (SerialPort serialPort : getSerialPorts()) {
            serialList.add(serialPort.getSystemPortName() + " - " + serialPort.getDescriptivePortName());
        }
        cbSerialList.getItems().addAll(serialList);
        cbSerialList.getSelectionModel().selectFirst();

        String lastSerial = ConfigManager.get(keyLastSerial, null);
        if (lastSerial != null && serialList.contains(lastSerial)) {
            cbSerialList.setValue(lastSerial);
            LOG.info("恢复上次串口选择: " + lastSerial);
        } else if (!serialList.isEmpty()) {
            cbSerialList.getSelectionModel().selectFirst();
        }

    }

    private void onSerialPortClicked() {
        String selected = cbSerialList.getValue();
        if (selected != null) {
            ConfigManager.set(keyLastSerial, selected);
            LOG.info("保存上次串口选择: " + selected);
        }
        closeSelectSerial();
        openSelectSerial(cbSerialList, cbBautRateList);
    }


    private void initOpenSerialButtonAction() {
        btnOpenSerial.selectedProperty().addListener((observableValue, oldVal, newVal) -> {
            if (newVal) {
                openSelectSerial(cbSerialList, cbBautRateList);
            } else {
                closeSelectSerial();
            }
        });
    }

    private void initBautRateComboBoxAction() {
        // 监听波特率选中项变化事件
        cbBautRateList.getSelectionModel().selectedItemProperty().addListener((observable) -> {
            onBaudRateChanged();
        });
    }

    private void onBaudRateChanged() {
        if (comPort != null && comPort.isOpen()) {
            closeSelectSerial();
            openSelectSerial(cbSerialList, cbBautRateList);
        }
    }

    private void closeSelectSerial() {
        if (null != comPort) {
            comPort.closePort();
        }
        if (null != serialReadService) {
            serialReadService.cancel();
        }
    }

    private void initBautRateList() {
        List<Integer> baudRates = new ArrayList<>(Arrays.asList( // list可以增删改
                1200, 2400, 4800, 9600, 38400, 57600, 115200, 230400, 1500000, 2000000, 3000000));

        // 清空旧的 items 并添加
        cbBautRateList.getItems().clear();
        cbBautRateList.getItems().addAll(baudRates);

        // 默认选中 115200
        cbBautRateList.getSelectionModel().select(Integer.valueOf(1500000));

        LOG.info("波特率初始化完成：" + cbBautRateList.getItems());
    }

    private void refreshSerialList() {
        // 刷新前记录当前选中的串口（用于后续恢复）
        String currentSelected = cbSerialList.getValue();

        new TaskHandler<List<String>>().whenCall(() -> {
            List<String> serialList = new ArrayList<>();
            for (SerialPort serialPort : getSerialPorts()) {
                serialList.add(serialPort.getSystemPortName() + " - " + serialPort.getDescriptivePortName());
            }
            return serialList;
        }).andThen(val -> {
            LOG.info("读取完成" + val);
            cbSerialList.getItems().clear();
            cbSerialList.getItems().addAll(val);

            // 刷新后：如果之前有选中项且仍存在，则恢复选中；否则不自动选中
            if (currentSelected != null && val.contains(currentSelected)) {
                cbSerialList.setValue(currentSelected); // 恢复之前的选中项
            } else {
                // 首次加载或选中项已消失，可选：不自动选中任何项
                cbSerialList.getSelectionModel().clearSelection();
            }
        }).handle();
    }

    private static List<SerialPort> getSerialPorts() {
        return Arrays.asList(SerialPort.getCommPorts());
    }

    public void setSerialName(String serialName) {
        this.lbSerialName.setText(serialName);
    }
    public void setKeyLastSerial(String keyLastSerial) {
        this.keyLastSerial = keyLastSerial;
    }
}

File: src/main/java/indi/lt/serialtool/controller/SerialReceiveCtrl.kt
package indi.lt.serialtool.controller

import com.fazecast.jSerialComm.SerialPort
import indi.lt.serialtool.component.InlineCssRegexHighlighter
import indi.lt.serialtool.component.PromptInlineCssTextArea
import indi.lt.serialtool.component.SerialPortCombBox
import indi.lt.serialtool.component.SerialToggleButton
import indi.lt.serialtool.data.SerialPortSettings
import indi.lt.serialtool.service.SerialReadService
import indi.lt.serialtool.global.ConfigManager
import indi.lt.serialtool.utils.UIUtil
import javafx.application.Platform
import javafx.beans.Observable
import javafx.beans.value.ObservableValue
import javafx.collections.FXCollections
import javafx.fxml.FXML
import javafx.fxml.FXMLLoader
import javafx.fxml.Initializable
import javafx.scene.control.*
import javafx.scene.layout.BorderPane
import javafx.stage.FileChooser
import javafx.util.Callback
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger
import java.io.FileWriter
import java.io.IOException
import java.net.URL
import java.util.*

/**
 * 接收模式逻辑
 * @author Nonoas
 * @date 2025/8/22
 * @since 1.0.0
 */
class SerialReceiveCtrl : Initializable {
    private val logger: Logger = LogManager.getLogger(MainController::class.java)

    // === FXML 注入的组件 ===
    @FXML
    private lateinit var rootPane: BorderPane

    @FXML
    private lateinit var menuBar: MenuBar

    @FXML
    private lateinit var lbSerialName: Label

    @FXML
    private lateinit var cbSerialList: SerialPortCombBox

    @FXML
    private lateinit var cbBautRateList: ComboBox<Int>

    @FXML
    private lateinit var textAreaOrigin: PromptInlineCssTextArea

    @FXML
    private lateinit var btnOpenSerial: SerialToggleButton

    @FXML
    private lateinit var cbTimeDisplay: CheckBox

    @FXML
    private lateinit var tfKeyWord: TextField

    @FXML
    private lateinit var lbRecvBytes: Label

    @FXML
    private lateinit var btnMoreSettings: Button

    // === 内部变量 ===
    private var serialReadService: SerialReadService? = null
    private var highlighter: InlineCssRegexHighlighter? = null
    private var keyLastSerial: String? = null

    // 串口参数设置
    private var serialPortSettings: SerialPortSettings = SerialPortSettings.createDefault()

    // 串口参数设置对话框 key
    private val KEY_SERIAL_SETTINGS = "receiveModeSerialSettings"

    override fun initialize(url: URL?, resourceBundle: ResourceBundle?) {
        initSerialPortSettings()
        initBautRateList()
        registerSerialEvent()
    }

    /**
     * 初始化串口参数设置
     */
    private fun initSerialPortSettings() {
        // 从配置加载串口参数设置
        serialPortSettings = ConfigManager.get(KEY_SERIAL_SETTINGS, SerialPortSettings::class.java, SerialPortSettings.createDefault())

        // 应用设置到串口组件
        cbSerialList.setSerialPortSettings(serialPortSettings)
    }

    private fun registerSerialEvent() {
        initOpenSerialButtonAction()
        initBautRateComboBoxAction()
        initMoreSettingsButtonAction()
    }

    /**
     * 初始化更多设置按钮动作
     */
    private fun initMoreSettingsButtonAction() {
        btnMoreSettings.setOnAction {
            showMoreSettings()
        }
    }

    /**
     * 显示更多设置对话框
     */
    fun showMoreSettings() {
        try {
            // 加载对话框 FXML
            val loader = FXMLLoader(javaClass.getResource("/fxml/serial-settings-dialog.fxml"))
            val dialogPane = loader.load<DialogPane>()

            // 获取控制器
            val dialogCtrl = loader.getController<SerialSettingsDialogCtrl>()

            // 设置当前设置
            dialogCtrl.setSettings(serialPortSettings)

            // 创建对话框
            val dialog = Dialog<SerialPortSettings>()
            dialog.dialogPane = dialogPane
            dialog.title = "串口参数设置"
            dialog.headerText = "自定义串口参数"

            // 设置按钮
            dialog.dialogPane.buttonTypes.addAll(ButtonType.OK, ButtonType.CANCEL)

            // 处理 OK 按钮
            dialog.resultConverter = Callback<ButtonType, SerialPortSettings> { buttonType ->
                if (buttonType == ButtonType.OK) {
                    // 从 UI 更新设置
                    dialogCtrl.updateSettingsFromUI()
                    dialogCtrl.getSettings()
                } else {
                    null
                }
            }

            // 显示对话框并等待结果
            dialog.showAndWait().ifPresent { settings ->
                // 保存设置
                serialPortSettings = settings
                ConfigManager.set(KEY_SERIAL_SETTINGS, settings)

                // 应用设置到串口组件
                cbSerialList.setSerialPortSettings(settings)

                // 显示成功提示
                UIUtil.showToast("串口参数已更新")
            }

        } catch (e: Exception) {
            logger.error("显示串口设置对话框失败", e)
            UIUtil.showToast("打开设置对话框失败")
        }
    }

    /**
     * 恢复自动滚动
     */
    @FXML
    private fun restoreScrolling() {
        textAreaOrigin.isAutoScroll = true
    }

    @FXML
    private fun clearLogs() {
        textAreaOrigin.area.clear()
        serialReadService?.resetRecvBytesCount()
        lbRecvBytes.text = "0 B"
    }

    @FXML
    private fun saveOriginLogs() {
        logger.info("saveOriginLogs")

        val content = textAreaOrigin.text
        if (content.isNullOrEmpty()) {
            logger.info("没有日志内容可保存")
            return
        }

        val fileChooser = FileChooser().apply {
            title = "保存日志文件"
            extensionFilters.addAll(
                FileChooser.ExtensionFilter("文本文件", "*.txt"), FileChooser.ExtensionFilter("所有文件", "*.*")
            )
        }

        val file = fileChooser.showSaveDialog(textAreaOrigin.scene.window) ?: return

        try {
            FileWriter(file, false).use { writer ->
                writer.write(content)
                logger.info("日志已保存到: ${file.absolutePath}")
            }
        } catch (e: IOException) {
            logger.error("保存日志失败", e)
        }
    }

    @FXML
    private fun saveFilterLogs() {
        logger.info("saveFilterLogs")
    }

    fun initSerialComboBoxAction() {
        cbSerialList.init(
            keyLastSerial, {
                UIUtil.getSelectedInt(cbBautRateList, 115200)
            }, btnOpenSerial.selectedProperty(), SerialPort.TIMEOUT_READ_SEMI_BLOCKING
        )
    }

    private fun initOpenSerialButtonAction() {
        cbSerialList.disableProperty().bind(btnOpenSerial.disableProperty())
        btnOpenSerial.selectedProperty().addListener { _: ObservableValue<out Boolean>?, _: Boolean?, newVal: Boolean ->
            if (newVal) {
                btnOpenSerial.isDisable = true
                cbSerialList.openSelectedSerial()
            } else {
                closeSelectSerial()
            }
        }
        cbSerialList.setOnOpenSucceed {
            btnOpenSerial.isDisable = false
            highlighter = InlineCssRegexHighlighter(textAreaOrigin).apply {
                patternTextProperty().bind(tfKeyWord.textProperty())
            }
            serialReadService = SerialReadService(
                cbSerialList.selectedPort, textAreaOrigin, cbTimeDisplay.selectedProperty()
            ) { highlighter?.schedule() }.also {
                it.setOnRecvBytesChanged { bytes ->
                    Platform.runLater {
                        lbRecvBytes.text = formatBytes(bytes)
                    }
                }
                it.start()
            }
        }
        cbSerialList.setOnOpenFailed {
            btnOpenSerial.isDisable = false
            btnOpenSerial.isSelected = false
        }
    }

    private fun initBautRateComboBoxAction() {
        if (!cbBautRateList.items.isEmpty()) {
            cbBautRateList.selectionModel.selectFirst()
        }
        cbBautRateList.selectionModel.selectedItemProperty().addListener { _: Observable? ->
            onBaudRateChanged()
        }
    }

    private fun onBaudRateChanged() {
        if (!cbSerialList.isActive) {
            return
        }
        closeSelectSerial()
        cbSerialList.openSelectedSerial()
    }

    private fun closeSelectSerial() {
        cbSerialList.closeSelectSerial()
        serialReadService?.cancel()
    }

    private fun initBautRateList() {
        val baudRates = FXCollections.observableArrayList(
            1200, 2400, 4800, 9600, 38400, 57600, 115200, 230400, 1500000, 2000000, 3000000
        )

        cbBautRateList.items = baudRates
        Platform.runLater {
            cbBautRateList.selectionModel.select(1500000)
            cbBautRateList.value = 1500000
        }
    }

    fun setSerialName(serialName: String?) {
        lbSerialName.text = serialName
    }

    fun setKeyLastSerial(keyLastSerial: String?) {
        this.keyLastSerial = keyLastSerial
    }

    /**
     * 将字节数格式化为人类可读的字符串
     */
    private fun formatBytes(bytes: Long): String {
        return when {
            bytes < 1024 -> "$bytes B"
            bytes < 1024 * 1024 -> String.format("%.2f KB", bytes / 1024.0)
            bytes < 1024 * 1024 * 1024 -> String.format("%.2f MB", bytes / (1024.0 * 1024.0))
            else -> String.format("%.2f GB", bytes / (1024.0 * 1024.0 * 1024.0))
        }
    }
}